Shrohilla/kafka header bindings

shrohilla opened the pull request:
1. Added the support for the kafka headers for output bindings
2. Reconverting the object from string check if the required params exists than create the KafkaEventData
3. It doesn't require the unit test case as the class is internal class

shrohilla commented:
/azp run

TsuyoshiUshio commented on src/Microsoft.Azure.WebJobs.Extensions.Kafka/Output/KafkaProducerAsyncCollector.cs:
```diff
@@ -32,14 +34,60 @@ public Task AddAsync(T item, CancellationToken cancellationToken)
             return entity.SendAndCreateEntityIfNotExistsAsync(messageToSend, functionInstanceId, cancellationToken);
         }
 
+        private object convertToKafkaEventData(T item)
```
Naming convention. In case of C# method name will be start with upper case.  The other methods are the same.

TsuyoshiUshio commented on src/Microsoft.Azure.WebJobs.Extensions.Kafka/Output/KafkaProducerAsyncCollector.cs:
```diff
@@ -32,14 +34,60 @@ public Task AddAsync(T item, CancellationToken cancellationToken)
+
+        private object buildKafkaDataEvent(T item)
+        {
+            JObject dataObj = JObject.Parse(item.ToString());
```
You can use .FromObject method https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_Linq_JObject_FromObject.htm

shrohilla replied:
I tried this and it didn't work so moved back to Parse function only

TsuyoshiUshio commented on src/Microsoft.Azure.WebJobs.Extensions.Kafka/Output/KafkaProducerAsyncCollector.cs:
```diff
@@ -32,14 +34,60 @@ public Task AddAsync(T item, CancellationToken cancellationToken)
+            messageToSend.Timestamp = (DateTime)dataObj["Timestamp"];
+            messageToSend.Partition = (int)dataObj["Partition"];
+            JArray headerList = (JArray)dataObj["Headers"];
+            foreach ( JObject header in headerList) {
```
Nit unnecessary space.

TsuyoshiUshio reviewed (changes requested):
Some naming convention issue is there. However generally looks good. 
One thing I'd like to recommend is, write unit test. You can use internal scope for enabling unit testing with internalVisuleTo https://codesnippets.fesslersoft.de/use-the-assembly-internalsvisibleto-attribute/

shrohilla commented:
> Some naming convention issue is there. However generally looks good. One thing I'd like to recommend is, write unit test. You can use internal scope for enabling unit testing with internalVisuleTo https://codesnippets.fesslersoft.de/use-the-assembly-internalsvisibleto-attribute/
> 
> Automated testing seems to fail.

Hi @TsuyoshiUshio I had addressed all the comments shared and created the unit test cases also, please review the code

TsuyoshiUshio commented on test/Microsoft.Azure.WebJobs.Extensions.Kafka.UnitTests/output/KafkaProducerAsyncCollectorTest.cs:
```diff
@@ -0,0 +1,69 @@
+            asyncCollector.AddAsync(Encoding.UTF8.GetBytes(testValue), default);
+        }
+
+        private void buildMockData()
```
Start with upper case.

shrohilla replied:
Resolved

TsuyoshiUshio reviewed (commented):
The unit tests should include the testing for headers since you added headers feature. You can share the spec and prove it works with variation with writing unit tests.

shrohilla commented:
> The unit tests should include the testing for headers since you added headers feature. You can share the spec and prove it works with variation with writing unit tests.

@TsuyoshiUshio  -- Already added that in the test -- AddAsync_Item_Is_Of_KafkaEventData_Json_String_Types
It contains headers details also

TsuyoshiUshio commented:
Hi @shrohilla 
The test doesn't include assertion or validation using Mock. Am I missing something? Usually, test cases will be 
1. Setup
2. Run the target
3. Assertion/Validation
However, on your test case I can't find it. For example. https://github.com/Azure/azure-functions-kafka-extension/blob/dev/test/Microsoft.Azure.WebJobs.Extensions.Kafka.UnitTests/KafkaProducerFactoryTest.cs  For the validation of the Moq will be https://makolyte.com/moq-verifying-parameters-passed-to-a-mocked-method/#:~:text=When%20you%20need%20to%20verify%20that%20the%20code,expected%20with%20the%20specified%20parameters.%20Here%E2%80%99s%20an%20example.

shrohilla commented:
> Hi @shrohilla The test doesn't include assertion or validation using Moq. Am I missing something? Usually, test cases will be constructed by three sections.
> 
> 1. Setup
> 2. Run the target
> 3. Assertion/Validation
>    However, on your test case I can't find it. For example. https://github.com/Azure/azure-functions-kafka-extension/blob/dev/test/Microsoft.Azure.WebJobs.Extensions.Kafka.UnitTests/KafkaProducerFactoryTest.cs  For the validation of the Moq will be https://makolyte.com/moq-verifying-parameters-passed-to-a-mocked-method/#:~:text=When%20you%20need%20to%20verify%20that%20the%20code,expected%20with%20the%20specified%20parameters.%20Here%E2%80%99s%20an%20example.

@TsuyoshiUshio Added the few possible validations, will add this scenario in the E2E tests for language workers as a separate task.
